File: android/app/src/main/java/it/fabiodalez/incitta/ui/ScrollNavigation.kt
package it.fabiodalez.incitta.ui

/** Compose scroll deltas are positive towards the top of the page. */
internal class ScrollNavigation(private val threshold: Float) {
    private var movement = 0f

    fun scroll(delta: Float): Boolean? {
        if (!delta.isFinite() || delta == 0f) return null
        movement = if (kotlin.math.sign(movement) == kotlin.math.sign(delta)) movement + delta else delta
        if (kotlin.math.abs(movement) < threshold) return null
        val visible = movement > 0
        movement = 0f
        return visible
    }
}

File: resources/js/scroll-navigation.js
/** Accumulate intentional direction changes, ignoring tiny scroll jitter. */
export function navigationMovement(previous, delta, threshold = 16) {
    if (!Number.isFinite(delta) || delta === 0) return { movement: previous, visible: null };
    const movement = Math.sign(delta) === Math.sign(previous) ? previous + delta : delta;
    return Math.abs(movement) >= threshold
        ? { movement: 0, visible: movement < 0 }
        : { movement, visible: null };
}
